File: terminal_ticker/config/agent_models.py
"""文件用途：Agent 模型配置，规范化 provider、model、api_mode 和推理强度。"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

CODEX_PROVIDER = "codex"
OPENAI_PROVIDER = "openai"
CODEX_API_MODE = "codex_responses"
OPENAI_CHAT_API_MODE = "chat_completions"
DEFAULT_CODEX_BASE_URL = "https://chatgpt.com/backend-api/codex"
DEFAULT_CODEX_MODEL = "gpt-5.4-mini"
DEFAULT_OPENAI_MODEL = "gpt-4o"

SUPPORTED_AGENT_PROVIDERS = {CODEX_PROVIDER, OPENAI_PROVIDER}
SUPPORTED_API_MODES = {CODEX_API_MODE, OPENAI_CHAT_API_MODE}
SUPPORTED_REASONING_EFFORTS = {"low", "medium", "high", "xhigh"}

CODEX_MODEL_ALIASES = {
    "default": DEFAULT_CODEX_MODEL,
    "codex": DEFAULT_CODEX_MODEL,
    "fast": DEFAULT_CODEX_MODEL,
}


@dataclass(frozen=True)
class AgentModelProfile:
    """说明：封装一次 Agent 请求最终使用的模型配置。"""

    provider: str
    api_mode: str
    model: str
    reasoning_effort: str
    supports_reasoning: bool = True
    requires_account_id: bool = True


def normalize_provider(raw_value: Any) -> str:
    """说明：规范化并校验 Agent provider。"""
    if raw_value is None:
        return CODEX_PROVIDER
    if not isinstance(raw_value, str):
        raise ValueError("agent.provider must be a string")
    provider = raw_value.strip().lower()
    if not provider:
        return CODEX_PROVIDER
    if provider not in SUPPORTED_AGENT_PROVIDERS:
        supported = ", ".join(sorted(SUPPORTED_AGENT_PROVIDERS))
        raise ValueError(f"agent.provider must be one of: {supported}")
    return provider


def normalize_api_mode(provider: str, raw_value: Any = None) -> str:
    """说明：解析 provider 对应的 API 模式。"""
    if raw_value is None or raw_value == "":
        if provider == CODEX_PROVIDER:
            return CODEX_API_MODE
        if provider == OPENAI_PROVIDER:
            return OPENAI_CHAT_API_MODE
    if not isinstance(raw_value, str):
        raise ValueError("agent.api_mode must be a string")
    api_mode = raw_value.strip().lower()
    if not api_mode:
        return normalize_api_mode(provider)
    if provider == CODEX_PROVIDER and api_mode != CODEX_API_MODE:
        raise ValueError(f"agent.api_mode for codex must be {CODEX_API_MODE}")
    if provider == OPENAI_PROVIDER and api_mode != OPENAI_CHAT_API_MODE:
        raise ValueError(f"agent.api_mode for openai must be {OPENAI_CHAT_API_MODE}")
    if api_mode not in SUPPORTED_API_MODES:
        supported = ", ".join(sorted(SUPPORTED_API_MODES))
        raise ValueError(f"agent.api_mode must be one of: {supported}")
    return api_mode


def normalize_model(provider: str, raw_value: Any) -> str:
    """说明：规范化 provider 对应的模型名称。"""
    if raw_value is None:
        if provider == CODEX_PROVIDER:
            return DEFAULT_CODEX_MODEL
        if provider == OPENAI_PROVIDER:
            return DEFAULT_OPENAI_MODEL
        return ""
    if not isinstance(raw_value, str):
        raise ValueError("agent.model must be a string")
    model = raw_value.strip()
    if not model:
        if provider == CODEX_PROVIDER:
            return DEFAULT_CODEX_MODEL
        if provider == OPENAI_PROVIDER:
            return DEFAULT_OPENAI_MODEL
        return ""
    if provider == CODEX_PROVIDER:
        return CODEX_MODEL_ALIASES.get(model.lower(), model)
    return model


def normalize_reasoning_effort(raw_value: Any) -> str:
    """说明：规范化模型推理强度。"""
    if raw_value is None:
        return "medium"
    if not isinstance(raw_value, str):
        raise ValueError("agent.reasoning_effort must be a string")
    effort = raw_value.strip().lower()
    if not effort:
        return "medium"
    aliases = {"minimal": "low", "extra": "xhigh", "extra_high": "xhigh"}
    normalized = aliases.get(effort, effort)
    if normalized not in SUPPORTED_REASONING_EFFORTS:
        supported = ", ".join(sorted(SUPPORTED_REASONING_EFFORTS))
        raise ValueError(f"agent.reasoning_effort must be one of: {supported}")
    return normalized


def resolve_agent_model(config: Any) -> AgentModelProfile:
    """说明：把 AgentConfig 解析为最终模型配置。"""
    provider = normalize_provider(getattr(config, "provider", None))
    api_mode = normalize_api_mode(provider, getattr(config, "api_mode", None))
    model = normalize_model(provider, getattr(config, "model", None))
    reasoning_effort = normalize_reasoning_effort(getattr(config, "reasoning_effort", None))
    if provider == CODEX_PROVIDER:
        return AgentModelProfile(
            provider=provider,
            api_mode=api_mode,
            model=model,
            reasoning_effort=reasoning_effort,
        )
    if provider == OPENAI_PROVIDER:
        return AgentModelProfile(
            provider=provider,
            api_mode=api_mode,
            model=model,
            reasoning_effort=reasoning_effort,
            supports_reasoning=False,
            requires_account_id=False,
        )
    raise ValueError(f"Unsupported agent provider: {provider}")

File: tradex/config/agent_models.ts
export const CODEX_PROVIDER = "codex";
export const ANTHROPIC_PROVIDER = "anthropic";
// OpenAI Chat Completions compatible provider. Covers OpenAI itself plus any
// OpenAI-compatible backend (LiteLLM proxy, Ollama, vLLM, OpenRouter, etc.).
// Following pi's design, providers are keyed by wire-format, not by vendor:
// switching backends is just changing `base_url`, never the provider code.
export const OPENAI_PROVIDER = "openai";
export const CODEX_API_MODE = "codex_responses";
export const ANTHROPIC_MESSAGES_API_MODE = "anthropic_messages";
export const OPENAI_COMPLETIONS_API_MODE = "openai_completions";
export const DEFAULT_CODEX_BASE_URL = "https://chatgpt.com/backend-api/codex";
export const DEFAULT_OPENAI_BASE_URL = "https://api.openai.com/v1";
export const DEFAULT_CODEX_MODEL = "gpt-5.4-mini";
export const DEFAULT_ANTHROPIC_MODEL = "global.anthropic.claude-opus-4-6-v1";
export const DEFAULT_OPENAI_MODEL = "gpt-4o";

const SUPPORTED_AGENT_PROVIDERS = new Set([CODEX_PROVIDER, ANTHROPIC_PROVIDER, OPENAI_PROVIDER]);
const SUPPORTED_API_MODES = new Set([CODEX_API_MODE, ANTHROPIC_MESSAGES_API_MODE, OPENAI_COMPLETIONS_API_MODE]);
const SUPPORTED_REASONING_EFFORTS = new Set(["low", "medium", "high", "xhigh", "max"]);
const CODEX_MODEL_ALIASES = new Map([
  ["default", DEFAULT_CODEX_MODEL],
  ["codex", DEFAULT_CODEX_MODEL],
  ["fast", DEFAULT_CODEX_MODEL],
]);

export interface AgentModelProfile {
  provider: string;
  apiMode: string;
  model: string;
  reasoningEffort: string;
  supportsReasoning: boolean;
  requiresAccountId: boolean;
}

export function normalizeProvider(rawValue: unknown): string {
  if (rawValue === null || rawValue === undefined) return CODEX_PROVIDER;
  if (typeof rawValue !== "string") throw new Error("agent.provider must be a string");
  const provider = rawValue.trim().toLowerCase();
  if (!provider) return CODEX_PROVIDER;
  if (!SUPPORTED_AGENT_PROVIDERS.has(provider)) {
    throw new Error(`agent.provider must be one of: ${[...SUPPORTED_AGENT_PROVIDERS].sort().join(", ")}`);
  }
  return provider;
}

export function normalizeApiMode(provider: string, rawValue?: unknown): string {
  if (rawValue === null || rawValue === undefined || rawValue === "") {
    if (provider === CODEX_PROVIDER) return CODEX_API_MODE;
    if (provider === ANTHROPIC_PROVIDER) return ANTHROPIC_MESSAGES_API_MODE;
    if (provider === OPENAI_PROVIDER) return OPENAI_COMPLETIONS_API_MODE;
  }
  if (typeof rawValue !== "string") throw new Error("agent.api_mode must be a string");
  const apiMode = rawValue.trim().toLowerCase();
  if (!apiMode) return normalizeApiMode(provider);
  if (provider === CODEX_PROVIDER && apiMode !== CODEX_API_MODE) {
    throw new Error(`agent.api_mode for codex must be ${CODEX_API_MODE}`);
  }
  if (provider === ANTHROPIC_PROVIDER && apiMode !== ANTHROPIC_MESSAGES_API_MODE) {
    throw new Error(`agent.api_mode for anthropic must be ${ANTHROPIC_MESSAGES_API_MODE}`);
  }
  if (provider === OPENAI_PROVIDER && apiMode !== OPENAI_COMPLETIONS_API_MODE) {
    throw new Error(`agent.api_mode for openai must be ${OPENAI_COMPLETIONS_API_MODE}`);
  }
  if (!SUPPORTED_API_MODES.has(apiMode)) {
    throw new Error(`agent.api_mode must be one of: ${[...SUPPORTED_API_MODES].sort().join(", ")}`);
  }
  return apiMode;
}

export function normalizeModel(provider: string, rawValue: unknown): string {
  if (rawValue === null || rawValue === undefined) {
    if (provider === CODEX_PROVIDER) return DEFAULT_CODEX_MODEL;
    if (provider === ANTHROPIC_PROVIDER) return DEFAULT_ANTHROPIC_MODEL;
    if (provider === OPENAI_PROVIDER) return DEFAULT_OPENAI_MODEL;
    return "";
  }
  if (typeof rawValue !== "string") throw new Error("agent.model must be a string");
  const model = rawValue.trim();
  if (!model) return normalizeModel(provider, null);
  if (provider === CODEX_PROVIDER) return CODEX_MODEL_ALIASES.get(model.toLowerCase()) ?? model;
  return model;
}

export function normalizeReasoningEffort(rawValue: unknown): string {
  if (rawValue === null || rawValue === undefined) return "medium";
  if (typeof rawValue !== "string") throw new Error("agent.reasoning_effort must be a string");
  const effort = rawValue.trim().toLowerCase();
  if (!effort) return "medium";
  const normalized = new Map([
    ["minimal", "low"],
    ["extra", "xhigh"],
    ["extra_high", "xhigh"],
  ]).get(effort) ?? effort;
  if (!SUPPORTED_REASONING_EFFORTS.has(normalized)) {
    throw new Error(`agent.reasoning_effort must be one of: ${[...SUPPORTED_REASONING_EFFORTS].sort().join(", ")}`);
  }
  return normalized;
}

export function resolveAgentModel(config: {
  provider?: unknown;
  apiMode?: unknown;
  api_mode?: unknown;
  model?: unknown;
  reasoningEffort?: unknown;
  reasoning_effort?: unknown;
}): AgentModelProfile {
  const provider = normalizeProvider(config.provider);
  const apiMode = normalizeApiMode(provider, config.apiMode ?? config.api_mode);
  const model = normalizeModel(provider, config.model);
  const reasoningEffort = normalizeReasoningEffort(config.reasoningEffort ?? config.reasoning_effort);
  if (provider === CODEX_PROVIDER) {
    return { provider, apiMode, model, reasoningEffort, supportsReasoning: true, requiresAccountId: true };
  }
  if (provider === ANTHROPIC_PROVIDER) {
    return { provider, apiMode, model, reasoningEffort, supportsReasoning: true, requiresAccountId: false };
  }
  if (provider === OPENAI_PROVIDER) {
    // reasoning is intentionally disabled for the first cut: OpenAI/o-series and
    // OpenAI-compatible proxies expose reasoning params in incompatible shapes.
    // Basic chat + tool-calling works without it; revisit per-backend later.
    return { provider, apiMode, model, reasoningEffort, supportsReasoning: false, requiresAccountId: false };
  }
  throw new Error(`Unsupported agent provider: ${provider}`);
}
